crates/uv-auth/src/store.rs

```diff
@@ -75,41 +20,64 @@ impl Default for AuthenticationStore {
 impl AuthenticationStore {
     pub fn new() -> Self {
         Self {
-            credentials: Mutex::new(HashMap::new()),
+            store: Mutex::new(HashMap::new()),
         }
     }
 
-    pub fn get(&self, url: &Url) -> Option<Option<Credential>> {
+    /// Retrieve stored credentials for a URL, if any.
+    pub fn get(&self, url: &Url) -> Option<Arc<Credentials>> {
         let netloc = NetLoc::from(url);
-        let credentials = self.credentials.lock().unwrap();
-        credentials.get(&netloc).cloned()
+        let store = self.store.lock().unwrap();
+        if let Some(fetched) = store.get(&netloc) {
+            fetched.clone()
+        } else {
+            None
+        }
     }
 
-    pub fn set(&self, url: &Url, auth: Option<Credential>) {
+    /// Set the stored credentials for a URL.
+    pub fn set(&self, url: &Url, credentials: Credentials) {
         let netloc = NetLoc::from(url);
-        let mut credentials = self.credentials.lock().unwrap();
-        credentials.insert(netloc, auth);
+        let mut store = self.store.lock().unwrap();
+        store.insert(netloc, Some(Arc::new(credentials)));
     }
 
-    /// Store in-URL credentials for future use.
-    pub fn save_from_url(&self, url: &Url) {
+    /// Populate the store with credentials on a URL, if there are any.
+    ///
+    /// If there are no credentials on the URL, the store will not be updated.
+    ///
+    /// Returns `true` if the store was updated.
+    pub fn set_from_url(&self, url: &Url) -> bool {
         let netloc = NetLoc::from(url);
-        let mut credentials = self.credentials.lock().unwrap();
-        if url.username().is_empty() {
-            // No credentials to save
-            return;
+        let mut store = self.store.lock().unwrap();
+        if let Some(credentials) = Credentials::from_url(url) {
+            store.insert(netloc, Some(Arc::new(credentials)));
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Returns `true` if we do not have credentials for the given URL and we not
+    /// have attempted to fetch them before.
+    pub fn should_attempt_fetch(&self, url: &Url) -> bool {
+        let netloc = NetLoc::from(url);
+        let store = self.store.lock().unwrap();
+        store.get(&netloc).is_none()
+    }
+
+    /// Track that fetching credentials for the given URL was attempted.
+    ///
+    /// If already set or if credentials have been populated for this URL, returns `false`.
+    pub fn set_fetch_attempted(&self, url: &Url) -> bool {
+        let netloc = NetLoc::from(url);
+        let mut store = self.store.lock().unwrap();
+        if store.get(&netloc).is_none() {
+            store.insert(netloc, None);
```

Comment:
I'm having trouble following this. If `get` returns `None`, we insert `None`?

Reply:
Is that because missing credentials would return `Some(None)`?

I'm wondering if we should use a custom enum here for missing credentials, instead of `Option<Credentials>`.

Reply:
Yes the latter `Some(None)`. An enum sounds nice although I might want to tackle that separately. I've already moved all the nested option logic _inside_ this struct in this pull request rather than having it be a part of its public API.

Reply:
I can stack it on top of this before merging though.